File: python-uv/src/puzzle/beginner/q07.py
#
# YYYYMMDD形式の日付を、2進数→逆にする→10進数としたとき、元の値と同じになるものは？
#
# /// script
# dependencies = [
#   "python-dateutil",
# ]
# ///
from datetime import date, datetime

# https://github.com/astral-sh/uv/issues/12030
from dateutil.rrule import DAILY, rrule

begin = datetime.now().microsecond / 1000

from_date = date(1964, 10, 10)
end_date = date(2020, 7, 24)

for d in rrule(DAILY, dtstart=from_date, until=end_date):
  base = int(d.strftime("%Y%m%d"))
  b02 = format(base, "b")[::-1]
  transformed = int(b02, 2)

  if base == transformed:
    print(f"a07. = {d.strftime('%Y-%m-%d')}")

finish = datetime.now().microsecond / 1000
print(f"{finish - begin}ms")

File: script/php/src/puzzle/beginner/q07.php
<?php

$begin = microtime(true);

$from = new DateTime('1964-10-10');
$end = new DateTime('2020-07-24');

while ($from <= $end) {
  $base = (int) $from->format('Ymd');
  $b02 = strrev(decbin($base));
  $transformed = bindec($b02);

  if ($base === $transformed) {
    echo 'a07. = '.$from->format('Y-m-d')."\n";
  }

  $from->modify('+1 day');
}

$finish = microtime(true);
echo (($finish - $begin) * 1000)."ms\n";
